[SPARK-52709][SQL] Fix parsing of  STRUCT<>

ManosGEM opened the pull request:
### What changes were proposed in this pull request?
This PR fixes an issue in Spark SQL's parser where empty or nested STRUCT<> types cause incorrect parenthesis tracking and parsing failures. Previously, the parser increased the parenthesis depth counter upon encountering the keyword STRUCT. Due to the operator precedence in the lexer (e.g., NEQ is matched before LT), a construct like STRUCT<> could incorrectly be tokenized as STRUCT and NEQ. This caused the parser to increase the nesting counter without ever decreasing it, eventually resulting in a syntax error. 
For example, the following valid query fails under the current logic:
`SELECT cast(null as STRUCT<>), 2 >> 1;`

To fix this, we adjusted the definition of the NEQ token in the SQL lexer so that it no longer matches <> when used in a complex data type. This ensures that the parser correctly interprets the angle brackets as part of a type specification rather than as a comparison operator.

### Why are the changes needed?
bug fix.
By modifying the NEQ token rule to avoid incorrectly matching <> in this context, we ensure that:
-     Empty STRUCT types like STRUCT<> are parsed correctly.
-     Nested and complex STRUCT types are supported without breaking parsing logic.
-     Queries with complex data types and bitwise operations (e.g., >>), are no longer broken due to incorrect token handling.

This change improves Spark SQL’s compatibility with standard SQL syntax and user expectations.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
A new test case has been added to PlanParserSuite.scala to specifically verify that queries containing CAST(null AS STRUCT<>), nested STRUCT<> (if applicable), and the >> operator now parse successfully into the expected logical plan.

### Was this patch authored or co-authored using generative AI tooling?
No

Closes SPARK-52709     
Reported by : @mihailom-db

ManosGEM committed:
[SPARK-52709][SQL] Fix parsing of STRUCT<> when followed by operators

This commit addresses a parsing issue where the `STRUCT<>` data type
was incorrectly handled, especially when appearing before operators
like the bitwise shift (`>>`).

The problem stemmed from the parser misinterpreting the angle brackets
of `STRUCT<>` as the 'not equal' operator (`<>`), leading to syntax errors.
The fix ensures `STRUCT<>` is correctly recognized as a data type.

A new test case in `PlanParserSuite.scala` confirms that queries
with `CAST(null AS STRUCT<>)`, nested structs, and `>>` now parse correctly.

mihailomilosevic2001 reviewed (commented):
@ManosGEM Thanks for working on this issue. I have left a few comments on the PR. Some of them are just to make code more durable to future errors of this type. Please feel free to ping me for another round of review when you go through them.

ManosGEM commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:
```diff
@@ -1979,4 +1979,49 @@ class PlanParserSuite extends AnalysisTest {
+
+  test("SPARK-52709: Parsing of empty and nested Struct<> followed by shiftRight operator") {
+    import org.apache.spark.sql.types.{NullType, StructType, IntegerType}
+    import org.apache.spark.sql.catalyst.parser.CatalystSqlParser
```
Yes, of course that is the way to go, but I saw some other tests on the file having imports likewise. 
I will move them.

ManosGEM commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:
```diff
@@ -1979,4 +1979,49 @@ class PlanParserSuite extends AnalysisTest {
     assert(unresolvedRelation2.isStreaming)
   }
+
+  test("SPARK-52709: Parsing of empty and nested Struct<> followed by shiftRight operator") {
```
I will try my best to rewrite it in the manner you propose.

ManosGEM commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:
```diff
@@ -1979,4 +1979,49 @@ class PlanParserSuite extends AnalysisTest {
+      """
+        |SELECT
+        |  CAST(null AS STRUCT<>),
+        |  CAST(null AS STRUCT<a: STRUCT<b: INT>>),
```
I will do.Thought of this as well, but then I thought I should keep it to the specific bug you posted.

ManosGEM commented on sql/api/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseLexer.g4:
```diff
@@ -518,7 +518,7 @@ ZONE: 'ZONE';
 EQ  : '=' | '==';
 NSEQ: '<=>';
-NEQ : '<>';
+NEQ : '<>' {complex_type_level_counter == 0}?;
```
Didn't try that with a shiftright operator afterwards. This itself returns null. But with the shiftright we got the bug allover again. Will keep working on the issue.

mihailom-db replied:
Mhm, you are right, could we just then remove NEQ from parser rules for Struct.

mihailom-db commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:
```diff
@@ -1979,4 +1979,49 @@ class PlanParserSuite extends AnalysisTest {
+    val sql =
+      """
+        |SELECT
+        |  CAST(null AS STRUCT<>),
```
I would say `SparkSqlParserSuite` should be used to test the behaviour we are fixing, and the `DataTypeParserSuite` should just be aware of empty structs. If there is no test in the file for empty structs, please add it.

ManosGEM replied:
In `DataTypeParserSuite` there is already a check for empty structs.

mihailom-db commented:
nit: We usually like to name the PR the same way the ticket is named on Jira. Could you please align the PR name with Jira?

ManosGEM commented:
> nit: We usually like to name the PR the same way the ticket is named on Jira. Could you please align the PR name with Jira?

remove the word "null" then? Because the only other difference is the [SQL] tag that I added according to the guidelines.

ManosGEM committed:
SPARK-52709: Modified STRUCT parser and add more tests for complex data types.

This commit follows SPARK-52709 by:

- Removing  from  parsing rules.

- Relocating relevant parser tests from  to .

- Refactoring the test setup for complex data types into a reusable helper function.

- Adding comprehensive tests for valid , nested , ,  and their combinations to .

- Adding negative tests for invalid empty  and  types to confirm correct  behavior.

ManosGEM committed:
Merge branch 'apache:master' into SPARK-52709-fix-empty-struct-parsing

mihailomilosevic2001 reviewed (approved):
LGTM, @MaxGekk @cloud-fan Could you please have a last review/merge.

cloud-fan commented:
good catch! merging to master/4.0!

cloud-fan commented:
@ManosGEM Can you open a branch 4.0 backport PR? thanks!

ManosGEM commented:
@cloud-fan  Everything should be the same in the PR but open it on the branch-4.0 ? Sorry but this is actually my first PR on Spark.

yadavay-amzn commented:
I've submitted a new fix in #55285 that addresses this at the parser level as @cloud-fan suggested. Instead of modifying the NEQ lexer rule, the parser decrements `complex_type_level_counter` when it matches `STRUCT NEQ` in the `dataType` rule. This avoids the regression with `ARRAY(col1 <> col2)`.
